src/main/java/com/example/dog/dao/JdbcDogDAO.java

```diff
@@ -48,60 +48,78 @@ public Dog get(int id) {
 
     @Override
     public Dog add(Dog dog) {
+        Connection conn = null;
         PreparedStatement statement;
-        try (Connection conn = dataSource.getConnection()) {
+        try {
+            conn = dataSource.getConnection();
+            conn.setAutoCommit(false);
             statement = conn.prepareStatement(SQL_INSERT_DOG, Statement.RETURN_GENERATED_KEYS);
             statement.setString(1, dog.getName());
             statement.setInt(2, dog.getHeight());
             statement.setInt(3, dog.getWeight());
             statement.setString(4, getDateString(dog.getDateOfBirth()));
             statement.executeUpdate();
+            conn.commit();
             try (ResultSet keys = statement.getGeneratedKeys()) {
                 if (keys.next()) {
                     dog.setId(keys.getInt(1));
                 }
             }
         } catch (SQLException e) {
-            throw new RuntimeException(e);
+            handleSQLException(conn, e);
+        } finally {
+            closeConnection(conn);
         }
         return dog;
     }
 
     @Override
     public Dog update(int id, Dog dog) {
+        Connection conn = null;
         PreparedStatement statement;
-        Dog updatedDog;
-        try (Connection conn = dataSource.getConnection()) {
+        Dog updatedDog = null;
+        try {
+            conn = dataSource.getConnection();
+            conn.setAutoCommit(false);
             statement = conn.prepareStatement(SQL_UPDATE_DOG);
             statement.setString(1, dog.getName());
             statement.setInt(2, dog.getHeight());
             statement.setInt(3, dog.getWeight());
             statement.setString(4, getDateString(dog.getDateOfBirth()));
             statement.setInt(5, id);
             int affectedRows = statement.executeUpdate();
+            conn.commit();
             if (affectedRows == 0) {
                 throw new ObjectNotFoundException("Dog with id=" + id + " was not found!");
             }
             updatedDog = get(id);
         } catch (SQLException e) {
-            throw new RuntimeException(e);
+            handleSQLException(conn, e);
```

Comment:
Will `finally` work you if you throw an exception from here?

Reply:
I'd say - yes, even if `catch` block throws an exception